### PR TITLE
Start scheduler on specified tasks

### LINKED ISSUE #4215: Run from any task(s)
With spawn-on-demand we can trigger a flow from any given task(s).

We can already do this within a running workflow.

As a way to start up a not-running workflow it will:
- supersede checkpoint restart (already removed)

and,

- provide a better option than warm start from a cycle point (in most cases)

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #4215

Start up with specified tasks ready to run. 
- replaces Cylc 7 checkpoint restart
- warm start anywhere in the graph  - not just at a cycle point

```console
# trigger foo.3 and bar.3 to start workflow blah/run1:
$ cylc play -t foo.3 -t bar.3 blah/run1
```
Naturally you can if you want trigger a side-branch that does not lead to the full graph being executed.

**Bonus:** :tada: ditch nasty ConditionalSimplifier code for pre-initial dependencies.  Why rewrite conditional expressions to ignore pre-initial triggers, when you can just set pre-initial prerequisites to satisfied? 
 
**Change in behaviour**: abort if scheduler run options are not consistent with type of start (new run vs restart), instead of ignoring the options and printing a warning message. Because: presence of such options probably means the user intended to do a different start than the run directory state indicates, and it is easy to miss the warning and assume that you did in fact do the intended kind of start. 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] No dependency changes
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [ ] (master branch) I have opened a documentation PR at cylc/cylc-doc/pull/XXXX.

### CODE REVIEW

**Comment by hjoliver** [comment]
Expecting a bunch of functional test fails to the change (/massive simplification) in pre-initial handling. It doesn't make any difference to how the workflow runs, but for reference tests affected tasks report now report `triggered of [<pre-initial deps>]` instead of `triggered off []`.

**Comment by hjoliver** [comment]
56 tests to fix, brilliant.

**Comment by hjoliver** [comment]
(Note to self: what's going on with `tests/f/cyclers/02-monthly/reference.log` (2nd trigger line) ? )

File: cylc/flow/listify.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+
+REC_CONDITIONALS = re.compile("([&|()])")
+
+
+def listify(message):
```
Comment (hjoliver) [inline cylc/flow/listify.py:23]: This function is all that remains of the `conditional_simplifier` module. Listification is still used in graph parsing.

File: cylc/flow/prerequisite.py [hunk 2]
```diff
@@ -89,19 +84,20 @@ def add(self, name, point, output, pre_initial=False):
...
+            pre_initial (bool): this is a pre-initial dependency.
 
         """
         message = (name, str(point), output)
 
-        # Add a new prerequisite message in an UNSATISFIED state.
-        self.satisfied[message] = self.DEP_STATE_UNSATISFIED
+        # Add a new prerequisite as satisfied if pre-initial, else unsatisfied.
+        if pre_initial:
+            self.satisfied[message] = self.DEP_STATE_SATISFIED
```
Comment (hjoliver) [inline cylc/flow/prerequisite.py:94]: This replaces the whole `ConditionalSimplifier` mess.

File: tests/functional/cylc-play/00-cycle-point-opts.t [hunk 3]
```diff
@@ -1,58 +0,0 @@
...
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <http://www.gnu.org/licenses/>.
-# -----------------------------------------------------------------------------
-# Test that invalid cycle point cli options for `cylc play` on
-# start vs restart are ignored
```
Comment (hjoliver) [inline tests/functional/cylc-play/00-cycle-point-opts.t:19]: These functional tests replaced by scheduler unit tests.

**Comment by hjoliver** [comment]
@kinow - assigning you for review at this end, but it can wait till next week if we're still busy with the table view etc.

@oliver-sanders - can you review or reassign to another UK team member

@dpmatthews - you might want to approve the "change in behaviour" with regard to `cylc play` options

**Review by oliver-sanders — COMMENTED** [review]
Code looks good, some nice simplification.

File: tests/functional/hold-release/17-hold-after-point/reference.log [hunk 4]
```diff
@@ -1,5 +1,5 @@
-2015-06-03T11:33:34Z INFO - Initial point: 20100101T0000Z
-2015-06-03T11:33:34Z INFO - Final point: 20100110T0000Z
-2015-06-03T11:33:34Z INFO - [stopper.20100101T0000Z] -triggered off []
-2015-06-03T11:33:34Z INFO - [foo.20100101T0000Z] -triggered off []
-2015-06-03T11:33:42Z INFO - [foo.20100102T0000Z] -triggered off ['foo.20100101T0000Z']
+Initial point: 20100101T0000Z
+Final point: 20100110T0000Z
+[stopper.20100101T0000Z] -triggered off []
+[foo.20100101T0000Z] -triggered off ['foo.20091231T0000Z']
```
Comment (oliver-sanders) [inline tests/functional/hold-release/17-hold-after-point/reference.log:4]: Just to check, the reference logs now list pre-initial tasks as `triggered off` values?

Comment (hjoliver) [reply tests/functional/hold-release/17-hold-after-point/reference.log:4]: Yes. Which is fine, as reference logs aren’t user facing. And (arguably at least) it’s just as valid as ignoring them.

Comment (oliver-sanders) [reply tests/functional/hold-release/17-hold-after-point/reference.log:4]: Happy with this, makes more sense IMO.

File: cylc/flow/scheduler.py [hunk 5]
```diff
@@ -1847,28 +1855,26 @@ def _get_events_conf(self, key, default=None):
         return self.workflow_event_handler.get_events_conf(
             self.config, key, default)
 
-    def process_cycle_point_opts(self) -> None:
-        """Check the values of --icp, --fcp, --startcp, --stopcp.
+    def _check_startup_opts(self) -> None:
+        """Abort if "cylc play" options are not consist with type of start.
```
Comment (kinow) [inline cylc/flow/scheduler.py:1859]: s/consist/consistent?

**Review by kinow — APPROVED** [review]
@hjoliver glanced at the code changes, ignoring most of the tests (you really had to update a lot of tests! :clap: ). Found no issues, just a typo I think.

Tested using Cylc UI and a modified `five` to play with integer points. Started at different tasks and cycle points (e.g. `prep.1`, `foo.2`, `foo.10`, etc), all looking good!

Tested with an invalid task, and got a nice warning message :heavy_check_mark: 

```bash
(venv) kinow@ranma:~/cylc-run/five$ cylc play --no-detach -t 404.2 five

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T17:09:41+12:00 INFO - Start task: ['404.2']
2021-05-24T17:09:41+12:00 WARNING - No matching tasks found: 404.2
2021-05-24T17:09:41+12:00 INFO - Scheduler: url=tcp://ranma:43090/ pid=7506
2021-05-24T17:09:41+12:00 INFO - Workflow publisher: url=tcp://ranma:43077
2021-05-24T17:09:41+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T17:09:41+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T17:09:41+12:00 INFO - Run mode: live
2021-05-24T17:09:41+12:00 INFO - Initial point: 1
2021-05-24T17:09:41+12:00 INFO - Final point: None
2021-05-24T17:09:41+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T17:09:42+12:00 INFO - DONE
```

And a different cycle point class and another useful message! :heavy_check_mark: 

```bash
(venv) kinow@ranma:~/cylc-run/five$ cylc play --no-detach -t foo.20130808T00 five

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T17:11:07+12:00 INFO - Start task: ['foo.20130808T00']
2021-05-24T17:11:07+12:00 WARNING - No matching tasks found: foo.20130808T00 (Incompatible value for <class
	'cylc.flow.cycling.integer.IntegerPoint'>: 20130808T00: invalid literal for int() with base 10: '20130808T00')
2021-05-24T17:11:07+12:00 INFO - Scheduler: url=tcp://ranma:43008/ pid=7544
2021-05-24T17:11:07+12:00 INFO - Workflow publisher: url=tcp://ranma:43035
2021-05-24T17:11:07+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T17:11:07+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T17:11:07+12:00 INFO - Run mode: live
2021-05-24T17:11:07+12:00 INFO - Initial point: 1
2021-05-24T17:11:07+12:00 INFO - Final point: None
2021-05-24T17:11:07+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T17:11:08+12:00 INFO - DONE
```

Then just because I was focusing on a more functional test, tested with an invalid cycle point, again a helpful message. :heavy_check_mark: 

```bash
(venv) kinow@ranma:~/cylc-run/five$ cylc play --no-detach -t foo.#### five

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T17:12:13+12:00 INFO - Start task: ['foo.####']
2021-05-24T17:12:13+12:00 WARNING - No matching tasks found: foo.#### (Incompatible value for <class 'cylc.flow.cycling.integer.IntegerPoint'>:
	####: invalid literal for int() with base 10: '####')
2021-05-24T17:12:13+12:00 INFO - Scheduler: url=tcp://ranma:43081/ pid=7560
2021-05-24T17:12:13+12:00 INFO - Workflow publisher: url=tcp://ranma:43091
2021-05-24T17:12:13+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T17:12:13+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T17:12:13+12:00 INFO - Run mode: live
2021-05-24T17:12:13+12:00 INFO - Initial point: 1
2021-05-24T17:12:13+12:00 INFO - Final point: None
2021-05-24T17:12:13+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T17:12:14+12:00 INFO - DONE
```

What about a negative number? Helpful warning followed by a traceback. Not sure if an error? :question: 

```bash
(venv) kinow@ranma:~/cylc-run/five$ cylc play --no-detach -t foo.-1 five

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T17:13:08+12:00 INFO - Start task: ['foo.-1']
2021-05-24T17:13:08+12:00 WARNING - Invalid cycle point for task: foo, -1
2021-05-24T17:13:08+12:00 INFO - Scheduler: url=tcp://ranma:43014/ pid=7588
2021-05-24T17:13:08+12:00 INFO - Workflow publisher: url=tcp://ranma:43049
2021-05-24T17:13:08+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T17:13:08+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T17:13:08+12:00 INFO - Run mode: live
2021-05-24T17:13:08+12:00 INFO - Initial point: 1
2021-05-24T17:13:08+12:00 INFO - Final point: None
2021-05-24T17:13:08+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T17:13:09+12:00 INFO - DONE
Exception ignored in: <function BaseEventLoop.__del__ at 0x7f1aa52c1940>
Traceback (most recent call last):
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/base_events.py", line 656, in __del__
    self.close()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/unix_events.py", line 58, in close
    super().close()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 92, in close
    self._close_self_pipe()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 99, in _close_self_pipe
    self._remove_reader(self._ssock.fileno())
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 274, in _remove_reader
    key = self._selector.get_key(fd)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 190, in get_key
    return mapping[fileobj]
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 71, in __getitem__
    fd = self._selector._fileobj_lookup(fileobj)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 225, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 42, in _fileobj_to_fd
    raise ValueError("Invalid file descriptor: {}".format(fd))
ValueError: Invalid file descriptor: -1
```

What if I use the Japanese keyboard? 

```bash
(venv) kinow@ranma:~/cylc-run/five$ cylc play --no-detach -t foo.5 -t prep.1 five

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T17:18:57+12:00 INFO - Start task: ['foo.5', 'prep.1']
2021-05-24T17:18:57+12:00 INFO - (no parent) spawned foo.5 h
2021-05-24T17:18:57+12:00 INFO - setting foo.5 ready to run
2021-05-24T17:18:57+12:00 INFO - (no parent) spawned prep.1 h
2021-05-24T17:18:57+12:00 INFO - setting prep.1 ready to run
2021-05-24T17:18:57+12:00 INFO - Scheduler: url=tcp://ranma:43007/ pid=8179
2021-05-24T17:18:57+12:00 INFO - Workflow publisher: url=tcp://ranma:43079
2021-05-24T17:18:57+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T17:18:57+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T17:18:57+12:00 INFO - Run mode: live
2021-05-24T17:18:57+12:00 INFO - Initial point: 1
2021-05-24T17:18:57+12:00 INFO - Final point: None
2021-05-24T17:18:57+12:00 INFO - [foo.5] -submit-num=01, host=ranma
2021-05-24T17:18:57+12:00 INFO - [foo.5] -triggered off []
2021-05-24T17:18:58+12:00 INFO - [foo.5] status=preparing: (internal)submitted at 2021-05-24T17:18:58+12:00  for job(01) flow(h)
2021-05-24T17:18:58+12:00 INFO - [foo.5] -health check settings: submission timeout=None, polling intervals=PT15M,...
2021-05-24T17:18:58+12:00 INFO - [client-command] graphql zmq://kinow@ranma:message
2021-05-24T17:18:59+12:00 INFO - [foo.5] status=submitted: (received)started at 2021-05-24T17:18:58+12:00  for job(01) flow(h)
2021-05-24T17:18:59+12:00 INFO - [foo.5] -health check settings: execution timeout=None, polling intervals=PT15M,...
2021-05-24T17:18:59+12:00 INFO - [client-command] pb_entire_workflow zmq://kinow@ranma:uiserver
2021-05-24T17:19:03+12:00 INFO - [client-command] graphql zmq://kinow@ranma:message
2021-05-24T17:19:04+12:00 INFO - [foo.5] status=running: (received)succeeded at 2021-05-24T17:19:03+12:00  for job(01) flow(h)
2021-05-24T17:19:04+12:00 INFO - (foo.5) spawned bar.5 flow(h)
2021-05-24T17:19:04+12:00 INFO - (foo.5) spawned foo.6 flow(h)
2021-05-24T17:19:05+12:00 INFO - [prep.1] -submit-num=01, host=ranma
2021-05-24T17:19:05+12:00 INFO - [prep.1] -triggered off []
...
```

Everything appears to be working OK. @hjoliver `cylc hold five foo.7` marked `foo.7` as `held`, but it was executed anyway. That is a known bug right?

Anyway, I watched until the flow caught up with the task that had already been held. Eventually the logs show ` Not spawning foo.5 (spawned in flow S)` :+1: @hjoliver the UI never removes the cycle point `5`. That's either a bug in the UI or (what I think is happening) the UI never got a delta pruned for the cycle point `5` or the UI got that delta, but then got another delta adding/updating something in the already gone cycle point `5` and decided to add it back. Will move it to a separate discussion.

Thanks!
Bruno

**Comment by kinow** [comment]
Ah, found the issue, the UI keeps receiving a delta for the cycle point `5`, even though the Flow Program detected that it had been already executed. @hjoliver should I file an issue for that here or UIS perhaps? See screenshot below:

image

It shows a fresh delta with the cycle points that are currently running in the flow. And the `5` cycle point also received a delta. Workflow used:

```bash
[scheduling]
  cycling mode = integer
  initial cycle point = 1
  #initial cycle point = 20130808T00
  #final cycle point = 20130808T12
  [[queues]]
     [[[default]]]
       limit = 1
  [[graph]]
    R1 = "prep => foo"
    P1 = "foo[-P1] => foo => bar"

[runtime]
  [[root]]
      script="sleep 5"
  [[prep]]
  [[foo]]
  [[bar]]
```

I just kept running `cylc play --no-detach -t ..... five` followed by `rm -rf .service share log` to test it.

**Comment by hjoliver** [comment]
> `cylc hold five foo.7` marked `foo.7` as held, but it was executed anyway. That is a known bug right?

Sort of, post-SOD task hold is being addressed in #3743

Currently hold acts only on already-spawned task proxies, so:
- if `foo.7` had already been spawned and is still in the waiting state, it would be marked as held, and would actually be held
- but if not spawned yet the "hold" command would be forgotten by the time it was spawned (so it would not be marked as held, and not be held)

Neither of which sounds like what you describe there?

**Comment by hjoliver** [comment]
> What about a negative number? Helpful warning followed by a traceback. Not sure if an error? question

Is that traceback repeatable?  I don't see it:

```console
$ cylc play --no-detach --start-task=foo.-1 hjo/run5

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T18:39:26+12:00 INFO - Start task: ['foo.-1']
2021-05-24T18:39:26+12:00 WARNING - Invalid cycle point for task: foo, -1
2021-05-24T18:39:26+12:00 INFO - Scheduler: url=tcp://niwa-1007823l:43003/ pid=4045
2021-05-24T18:39:26+12:00 INFO - Workflow publisher: url=tcp://niwa-1007823l:43094
2021-05-24T18:39:26+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T18:39:26+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T18:39:26+12:00 INFO - Run mode: live
2021-05-24T18:39:26+12:00 INFO - Initial point: 1
2021-05-24T18:39:26+12:00 INFO - Final point: None
2021-05-24T18:39:27+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T18:39:27+12:00 INFO - [client-command] pb_entire_workflow
        zmq://oliverh@niwa-1007823l:uiserver
2021-05-24T18:39:27+12:00 INFO - DONE
```

**Comment by kinow** [comment]
> > `cylc hold five foo.7` marked `foo.7` as held, but it was executed anyway. That is a known bug right?
> 
> Sort of, post-SOD task hold is being addressed in #3743
> 
> Currently hold acts only on already-spawned task proxies, so:
> 
>     * if `foo.7` had already been spawned and is still in the waiting state, it would be marked as held, and would actually be held
> 
>     * but if not spawned yet the "hold" command would be forgotten by the time it was spawned (so it would not be marked as held, and not be held)
> 
> 
> Neither of which sounds like what you describe there?

Yup, I remembered seeing a discussion about it somewhere, thanks for linking the issue :+1: 

> > What about a negative number? Helpful warning followed by a traceback. Not sure if an error? question
> 
> Is that traceback repeatable? I don't see it:
> 
> ```
> $ cylc play --no-detach --start-task=foo.-1 hjo/run5
> 
>                 Cylc Workflow Engine 8.0b2.dev
>                 Copyright (C) 2008-2021 NIWA
>                 & British Crown (Met Office) & Contributors
> 
> 2021-05-24T18:39:26+12:00 INFO - Start task: ['foo.-1']
> 2021-05-24T18:39:26+12:00 WARNING - Invalid cycle point for task: foo, -1
> 2021-05-24T18:39:26+12:00 INFO - Scheduler: url=tcp://niwa-1007823l:43003/ pid=4045
> 2021-05-24T18:39:26+12:00 INFO - Workflow publisher: url=tcp://niwa-1007823l:43094
> 2021-05-24T18:39:26+12:00 INFO - Run: (re)start=0 log=1
> 2021-05-24T18:39:26+12:00 INFO - Cylc version: 8.0b2.dev
> 2021-05-24T18:39:26+12:00 INFO - Run mode: live
> 2021-05-24T18:39:26+12:00 INFO - Initial point: 1
> 2021-05-24T18:39:26+12:00 INFO - Final point: None
> 2021-05-24T18:39:27+12:00 INFO - Workflow shutting down - AUTOMATIC
> 2021-05-24T18:39:27+12:00 INFO - [client-command] pb_entire_workflow
>         zmq://oliverh@niwa-1007823l:uiserver
> 2021-05-24T18:39:27+12:00 INFO - DONE
> ```

Oh, interesting! Not happening. I also got a similar traceback with a task name in Japanese, but then it didn't happen again. Strange, let me take a look again...

**Comment by hjoliver** [comment]
> Anyway, I watched until the flow caught up with the task that had already been held. Eventually the logs show Not spawning foo.5 (spawned in flow S)

Presumably you mean "caught up with the task that had already been **spawned** (and executed)"?  

> `cylc play --no-detach -t foo.5 -t prep.1 five`

This would be a weird way to start up this workflow, which would suggest probable "user error"! (Because triggering `prep.1` alone is sufficient to start the whole workflow from the beginning; starting up by triggering `foo.5` immediately as well means starting another part of the same flow out in front.) ... however we should of course test that it works as expected :+1: (so, well done :grin: )

I tried this myself and watched with `cylc tui`, and it all worked fine, eventually carrying on past cycle point 5.

From your cylc-ui result it sounds like a problem with the deltas, for this case.  I'll try it myself...

**Comment by hjoliver** [comment]
Yes, same result.  From what the UI shows, either we get no prune delta for `foo.5` or we keep getting an erroneous delta for it.  One for you @dwsutherland !

**Comment by kinow** [comment]
>Presumably you mean "caught up with the task that had already been spawned (and executed)"?

Yup! Not enough :coffee: sorry! :smile: 

>Yes, same result. From what the UI shows, either we get no prune delta for foo.5 or we keep getting an erroneous delta for it. One for you @dwsutherland !

:pray: I'm confirming now whether the pruned delta was received or not (should have done earlier, sorry).

---

@hjoliver if I execute this repeatedly, I get the error occasionally. Probably some concurrency issue and not related to this issue?

```bash
(venv) kinow@ranma:~/cylc-run/five$ rm -rf .service && cylc play --no-detach five -t foo.-1

                Cylc Workflow Engine 8.0b2.dev
                Copyright (C) 2008-2021 NIWA
                & British Crown (Met Office) & Contributors

2021-05-24T21:42:47+12:00 INFO - Start task: ['foo.-1']
2021-05-24T21:42:47+12:00 WARNING - Invalid cycle point for task: foo, -1
2021-05-24T21:42:47+12:00 INFO - Scheduler: url=tcp://ranma:43059/ pid=20316
2021-05-24T21:42:47+12:00 INFO - Workflow publisher: url=tcp://ranma:43099
2021-05-24T21:42:47+12:00 INFO - Run: (re)start=0 log=1
2021-05-24T21:42:47+12:00 INFO - Cylc version: 8.0b2.dev
2021-05-24T21:42:47+12:00 INFO - Run mode: live
2021-05-24T21:42:47+12:00 INFO - Initial point: 1
2021-05-24T21:42:47+12:00 INFO - Final point: None
2021-05-24T21:42:47+12:00 INFO - Workflow shutting down - AUTOMATIC
2021-05-24T21:42:48+12:00 INFO - DONE
Exception ignored in: <function BaseEventLoop.__del__ at 0x7ffb45d9b940>
Traceback (most recent call last):
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/base_events.py", line 656, in __del__
    self.close()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/unix_events.py", line 58, in close
    super().close()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 92, in close
    self._close_self_pipe()
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 99, in _close_self_pipe
    self._remove_reader(self._ssock.fileno())
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/asyncio/selector_events.py", line 274, in _remove_reader
    key = self._selector.get_key(fd)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 190, in get_key
    return mapping[fileobj]
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 71, in __getitem__
    fd = self._selector._fileobj_lookup(fileobj)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 225, in _fileobj_lookup
    return _fileobj_to_fd(fileobj)
  File "/home/kinow/Development/python/miniconda3/lib/python3.8/selectors.py", line 42, in _fileobj_to_fd
    raise ValueError("Invalid file descriptor: {}".format(fd))
ValueError: Invalid file descriptor: -1

```

**Review by oliver-sanders — APPROVED** [review]
Code looks good, tried and tested. Haven't hit the bug above, but also haven't hit Cylc as hard in my testing.

There is a slight disjoint with the data store, here's an example with a simple sequential graph:

```
P1 = a[-P1] => a
```

Starting the flow with start tasks `a.1` & `a.5`:

warm

Note that task `a.4` is displayed as succeeded because it is a pre-initial task, however, it is then reset to waiting => submitted => running => succeeded. The task `a.5` is then displayed in the waiting state despite having already run.

**Comment by kinow** [comment]
I can see the pruned delta for the job of `foo.5`.

image

And the pruned delta for the pruned task and the cycle point (root family pruned, then we prune the CP).

image

It looks like after that I'm getting an expired delta for `foo.5`?

image

Here's the list of deltas after grepping for kinow|five|5

https://gist.github.com/kinow/54e0b0796e4464dd2065defd4089f2aa

**Comment by hjoliver** [comment]
@oliver-sanders -

> Note that task a.4 is displayed as succeeded because it is a pre-initial task, however, it is then reset to waiting => submitted => running => succeeded. The task a.5 is then displayed in the waiting state despite having already run.

I believe that's because of the way the datastore currently infers n=1 task status. A similar thing happens on master if you cold-start the workflow and immediately trigger a reflow at `a.5` (the only difference being `a.5` gets a new flow label in that case, which means `a.5` can run again when the original flow - from `a.1` - reaches that point in the graph).

Here, the scheduler starts up with `a.5` in the pool and immediately sets its prerequisites satisfied, so it can run. `a.4` is then an n=1 task, and the datastore presumably infers that it must have succeeded in order for `a.5` to be running.

Later, when the flow from `a.1` reaches - and runs - `a.4` the datastore will see `a.5` as an upcoming waiting n=1 task (it doesn't know that `a.5` already ran in the same flow and won't actually run again).

IMO this makes sense and is OK as-is, so long as it is understood that the datastore currently infers n = 1 status and doesn't actually consult the DB to see if they already ran in the same or previous flows.

**Comment by hjoliver** [comment]
> It looks like after that I'm getting an expired delta for foo.5?

:thinking: there's no actual task expiry going on here, but (vague recollection...) we might be temporarily re-purposing the "expired" state for some other use.  I'll take a look tomorrow.

**Comment by dpmatthews** [comment]
> @dpmatthews - you might want to approve the "change in behaviour" with regard to `cylc play` options

Sounds OK. I want to have a play with this but I'm happy that we can deal with any issues as follow-ups.

**Comment by hjoliver** [comment]
@kinow in your asyncio traceback above the invalid `-1` file descriptor does look suspiciously like the cycle point of your start-up task `foo.-1`, which is weird.  Can you try to reproduce it with say `foo.-8` and see if that holds true?

**Comment by kinow** [comment]
Hmm, strange. Tried over 30 times with `foo.-8`, and no traceback. Tried again with `foo.-1` and nothing as well. :thinking:

**Comment by hjoliver** [comment]
Well thanks for trying!  I think we can assume that semi-irreproducible bug is not related to this PR.

**Pull request merged**